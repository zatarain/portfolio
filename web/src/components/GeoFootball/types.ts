export interface Station {
	id: number,
	name: string,
	slug: string,
	country: string,
	time_zone: string,
	latitude: number,
	longitude: number,
	location: string,
	info_en: string,
	info_es: string,
}

export interface Stadium {
	id: number,
	name: string,
	slug: string,
	team: string,
	latitude: number,
	longitude: number,
	location: string,
}
