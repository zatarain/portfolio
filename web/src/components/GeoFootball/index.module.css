.map {
	height: 81vh;
}

.popup {
	display: grid;
	grid-template-columns: 2fr 4fr;
}

.popup dt {
	display: inline-block;
	font-weight: bold;
	margin: 0;
	text-align: right;
}

.popup dd {
	display: inline-block;
	margin: 0 1ch;
}
