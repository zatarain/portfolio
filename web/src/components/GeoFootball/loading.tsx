import Image from 'next/image'

interface Properties {
	text: string,
}

const Loading = ({ text }: Properties) => {
	return (
		<>
			<Image src="https://i.gifer.com/VAyR.gif" alt={text} width={16} height={16} /> {text}
		</>
	)
}

export default Loading
