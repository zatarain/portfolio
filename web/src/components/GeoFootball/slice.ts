import type { Station } from './types'

export interface GeoFootballState {
	latitude: number,
	longitude: number,
}

const BASE_URL = process.env.API_URL || process.env.NEXT_PUBLIC_API_URL

async function GET(path: string) {
	return await fetch(`${BASE_URL}${path}`, {
		headers: {
			'Content-Type': 'application/json',
		},
		method: 'GET',
	})
}

async function POST(path: string, data: string): Promise<Response> {
	return await fetch(`${BASE_URL}${path}`, {
		body: data,
		headers: {
			'Content-Type': 'application/json',
		},
		method: 'POST',
	})
}

async function DELETE(path: string): Promise<Response> {
	return await fetch(`${BASE_URL}${path}`, {
		headers: {
			'Content-Type': 'application/json',
		},
		method: 'DELETE',
	})
}

export async function getStations(): Promise<Response> {
	return GET('/stations')
}

export async function saveStation(station: Station): Promise<Response> {
	return POST('/stations', JSON.stringify(station))
}

export async function deleteStation(id: number): Promise<Response> {
	return DELETE(`/stations/${id}`)
}

export function flag(country: string) {
	const points = country
		.toUpperCase()
		.split('')
		.map((char) => 127397 + char.charCodeAt(0));
	return String.fromCodePoint(...points);
}
