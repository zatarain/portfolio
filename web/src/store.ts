import { configureStore, ThunkAction, Action, createSlice } from '@reduxjs/toolkit'

export function makeStore() {
  const responsiveFlipper = createSlice({
    name: 'flipper',
    initialState: {
      responsive: false,
    },
    reducers: {
      flipResponsive: (state) => {
        state.responsive = !state.responsive
      }
    }
  })
  return configureStore({
    reducer: { flipper: responsiveFlipper.reducer },
  })
}

const store = makeStore()

export type AppState = ReturnType<typeof store.getState>

export type AppDispatch = typeof store.dispatch

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  AppState,
  unknown,
  Action<string>
>

export default store
