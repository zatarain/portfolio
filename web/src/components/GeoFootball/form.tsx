import { Marker, Popup, useMapEvents } from 'react-leaflet'
import { LeafletEvent, Icon, Marker as LeafletMarker } from 'leaflet'
import { useForm } from 'react-hook-form'
import { useState } from 'react'
import { Noto_Color_Emoji } from 'next/font/google'

import type { Station, GroupedStations } from './types'
import Loading from './loading'
import { flag, saveStation } from './slice'

import styles from './index.module.css'
import { useDispatch } from 'react-redux'

const emoji = Noto_Color_Emoji({ weight: '400', subsets: ['emoji'], preload: false })

const pin = new Icon({
	iconUrl: 'https://cdn-icons-png.flaticon.com/512/2684/2684860.png',
	iconSize: [24, 24],
	iconAnchor: [0, 24],
})

interface Properties {
	clusters: GroupedStations,
	setClusters: Function,
}


const MapForm = ({ clusters, setClusters }: Properties) => {
	const countries = Object.keys(clusters)

	const initialStation = {
		name: '',
		country: 'GB',
		time_zone: 'Europe/London',
		latitude: 51.478,
		longitude: 0,
		info_en: '',
	} as Station

	const [station, setStation] = useState(initialStation)
	const [marker, setMarker] = useState(null)

	const {
		register,
		handleSubmit,
		formState: { isSubmitting, errors },
		reset,
		setError,
	} = useForm({
		defaultValues: { ...initialStation },
	});

	useMapEvents({
		dblclick(event) {
			setStation({
				...station,
				latitude: event.latlng.lat,
				longitude: event.latlng.lng,
			})
			marker && (marker as LeafletMarker).openPopup()
		}
	})

	const save = async (data: object | undefined) => {
		const response = await saveStation({
			...data,
			latitude: station.latitude,
			longitude: station.longitude,
		} as Station)
		if (response.ok) {
			marker && (marker as LeafletMarker).closePopup()
			const record = await response.json() as Station
			setStation({ ...initialStation });
			const cluster = clusters[record.country] as Station[]

			setClusters({
				...clusters,
				[record.country]: [...cluster, record],
			})
			reset()
		} else if (response.status == 400) {
			const errors: { [field: string]: string } = await response.json()
			for (const [field, message] of Object.entries(errors)) {
				setError(field as keyof Station, { type: 'custom', message })
			}
		} else {
			const message = await response.text()
			console.error(`Unknown error: ${message}`)
		}

		return response
	}

	const afterAdd = (event: LeafletEvent) => setMarker(event.target)

	return (
		<Marker position={[station.latitude, station.longitude]} icon={pin} eventHandlers={{ add: afterAdd }}>
			<Popup>
				<form id="popup-add-form" className={`${styles.form}`} method="POST" onSubmit={handleSubmit(save)}>
					<h3>Add New Marker</h3>
					<label htmlFor="name">Name:</label>
					<input id="name" type="text" role="textbox" required {...register('name')} />
					<div className={styles.error}>{errors.name?.message}</div>
					<div className={styles.location}>
						<div>
							<label htmlFor="latitude">Latitude:</label>
							<input id="latitude" type="text" name="latitude" value={station.latitude.toFixed(5)} disabled />
						</div>
						<div>
							<label htmlFor="longitude">Longitude:</label>
							<input id="longitude" type="text" name="longitude" value={station.longitude.toFixed(5)} disabled />
						</div>
					</div>
					<div className={styles['country-time']}>
						<div className={styles.country}>
							<label htmlFor="country">Country:</label>
							<select id="country" className={emoji.className} {...register('country')}>
								{countries.map((country: string) =>
									<option className={emoji.className} key={country} value={country}>{flag(country)}</option>)}
							</select>
						</div>
						<div className={styles.time}>
							<label htmlFor="time-zone">Timezone:</label>
							<input id="time-zone" type="text" {...register('time_zone')} />
						</div>
					</div>
					<label htmlFor="info-en">Additional information:</label>
					<input id="info-en" type="text" {...register('info_en')} />
					<button type="submit" disabled={isSubmitting}>
						{isSubmitting ? <Loading text="Saving..." /> : 'Save'}
					</button>
				</form>
			</Popup>
		</Marker>
	)
}

MapForm.displayName = 'MapForm'

export default MapForm
