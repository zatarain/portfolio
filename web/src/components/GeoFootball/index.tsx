import type { Station } from './types'
import { MapContainer, TileLayer, Marker, Popup, useMapEvents } from 'react-leaflet'
import { Icon, Marker as LeafletMarker } from 'leaflet'
import MarkerClusterGroup from 'react-leaflet-cluster'
import { useState, useRef } from 'react'
import { useForm } from 'react-hook-form'
import { saveStation } from './slice'

import { Noto_Color_Emoji } from 'next/font/google'
import styles from './index.module.css'
import { stat } from 'fs'

const emoji = Noto_Color_Emoji({ weight: '400', subsets: ['emoji'], preload: false })

interface Properties {
	stationsByCountry: { [country: string]: Station[] },
}

function flag(country: string) {
	const points = country
		.toUpperCase()
		.split('')
		.map((char) => 127397 + char.charCodeAt(0));
	return String.fromCodePoint(...points);
}

const train = new Icon({
	iconUrl: 'https://cdn-icons-png.flaticon.com/512/1702/1702305.png',
	iconSize: [24, 24],
	iconAnchor: [12, 12],
})

const stadium = new Icon({
	iconUrl: 'https://cdn-icons-png.flaticon.com/512/1540/1540530.png',
	iconSize: [24, 24],
	iconAnchor: [12, 12],
})

const pin = new Icon({
	iconUrl: 'https://cdn-icons-png.flaticon.com/512/2684/2684860.png',
	iconSize: [24, 24],
	iconAnchor: [0, 24],
})

interface MapFormProperties {
	clusters: { [country: string]: Station[] },
	setClusters: Function,
}

interface LoadingProperties {
	text: string,
}

const Loading = ({ text }: LoadingProperties) => {
	return (
		<>
			<img src="https://i.gifer.com/VAyR.gif" alt={text} width={16} height={16} /> {text}
		</>
	)
}

const MapForm = ({ clusters, setClusters }: MapFormProperties) => {
	const countries = Object.keys(clusters)
	const initialStation = {
		name: '',
		slug: '',
		country: 'GB',
		time_zone: 'Europe/London',
		latitude: 51.478,
		longitude: 0,
	} as Station
	const [station, setStation] = useState(initialStation)

	const marker = useRef(null)
	const {
		register,
		handleSubmit,
		formState: { isSubmitting, errors },
		setError,
	} = useForm({
		defaultValues: { ...initialStation },
	});

	useMapEvents({
		dblclick(event) {
			setStation({
				...station,
				latitude: event.latlng.lat,
				longitude: event.latlng.lng,
			})

			if (marker.current) {
				(marker.current as LeafletMarker).openPopup()
			}
		}
	})

	const save = async (data: object) => {
		const response = await saveStation({
			...data,
			latitude: station.latitude,
			longitude: station.longitude,
		} as Station)
		if (response.ok) {
			if (marker.current) {
				const current = (marker.current as LeafletMarker)
				current.closePopup()
			}
			const record = await response.json() as Station
			console.log('Response:', record)
			setStation({ ...initialStation });
			const cluster = clusters[record.country] as Station[]
			cluster.push(record)

			setClusters({
				...clusters,
				[record.country]: [...cluster],
			})
		} else if (response.status == 400) {
			const errors: { [field: string]: string } = await response.json()
			for (const [field, message] of Object.entries(errors)) {
				setError(field, { type: 'custom', message })
			}
		} else {
			console.error('Unknown error:', await response.text())
		}

		return response
	}

	return (
		<Marker ref={marker} position={[station.latitude, station.longitude]} icon={pin}>
			<Popup>
				<form id="popup-add-form" className={`${styles.form}`} method="POST" onSubmit={handleSubmit(save)}>
					<h3>Add New Marker</h3>
					<label htmlFor="name">Name: </label>
					<input type="text" required {...register('name')} />
					<div className={styles.error}>{errors.name?.message}</div>
					<div className={styles.location}>
						<div>
							<label htmlFor="latitude">Latitude: </label>
							<input type="text" name="latitude" value={station.latitude.toFixed(5)} disabled />
						</div>
						<div>
							<label htmlFor="longitude">Longitude: </label>
							<input type="text" name="longitude" value={station.longitude.toFixed(5)} disabled />
						</div>
					</div>
					<label htmlFor="slug">Slug: </label>
					<input type="text" {...register('slug')} />
					<div className={styles['country-time']}>
						<div className={styles.country}>
							<label htmlFor="country">Country: </label>
							<select className={emoji.className} {...register('country')}>
								{countries.map((country) =>
									<option className={emoji.className} key={country} value={country}>{flag(country)}</option>)}
							</select>
						</div>
						<div className={styles.time}>
							<label htmlFor="time_zone">Timezone: </label>
							<input type="text" {...register('time_zone')} />
						</div>
					</div>
					<button type="submit" disabled={isSubmitting}>
						{isSubmitting ? <Loading text="Saving..." /> : 'Save'}
					</button>
				</form>
			</Popup>
		</Marker >
	)
}

const GeoFootball = ({ stationsByCountry }: Properties) => {
	const openStreetMaps = 'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png'
	const copyright = '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'

	const [clusters, setClusters] = useState(stationsByCountry)

	return (
		<MapContainer className={styles.map} center={[51.505, 0]} zoom={5} scrollWheelZoom doubleClickZoom={false}>
			<TileLayer attribution={copyright} url={openStreetMaps} />
			<MapForm clusters={clusters} setClusters={setClusters} />
			{Object.entries(clusters).map(([country, stations]) =>
				<MarkerClusterGroup key={country} chunkedLoading>
					{(stations as Station[]).map((station: Station) =>
						<Marker key={`train-${station.id}`} position={[station.latitude, station.longitude]} icon={train}>
							<Popup key={`train-popup-${station.id}`}>
								<div className={styles.popup}>
									<h3>{station.name}</h3>
									<dl>
										<dt>Country: </dt><dd>{country} <span className={emoji.className}>{flag(country)}</span></dd>
										<dt>Position: </dt><dd>{station.latitude.toFixed(5)}, {station.longitude.toFixed(5)}</dd>
										<dt>Timezone: </dt><dd>{station.time_zone}</dd>
									</dl>
									<p>{station.info_en}</p>
								</div>
							</Popup>
						</Marker>
					)}
				</MarkerClusterGroup>
			)}

		</MapContainer>
	)
}

export default GeoFootball
