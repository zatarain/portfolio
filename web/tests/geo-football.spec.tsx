import { Provider } from 'react-redux'
import store from '#store'
import { render } from '@testing-library/react'
import GeoFootballPage, { getServerSideProps } from '#pages/geo-football'
import { Station } from '#components/GeoFootball/types'

const testData = [
	{
		id: 1,
		name: 'Canary Wharf',
		slug: 'canary-wharf',
		country: 'GB',
		time_zone: 'Europe/London',
		latitude: 51.50361,
		longitude: -0.01861,
		info_en: 'London Underground Station',
	} as Station,
	{
		id: 2,
		name: 'Dublin Heuston',
		slug: 'dublin-heuston',
		country: 'IE',
		time_zone: 'Europe/Dublin',
		latitude: 53.34583,
		longitude: -6.29530,
		info_en: 'Irish Trains',
	} as Station,
]

const testClusters = {
	GB: [
		{
			id: 1,
			name: 'Canary Wharf',
			slug: 'canary-wharf',
			country: 'GB',
			time_zone: 'Europe/London',
			latitude: 51.50361,
			longitude: -0.01861,
			info_en: 'London Underground Station',
		} as Station,
	],
	IE: [
		{
			id: 2,
			name: 'Dublin Heuston',
			slug: 'dublin-heuston',
			country: 'IE',
			time_zone: 'Europe/Dublin',
			latitude: 53.34583,
			longitude: -6.29530,
			info_en: 'Irish Trains',
		} as Station,
	],
}

global.fetch = jest.fn(() =>
	Promise.resolve({
		ok: true,
		json: () => Promise.resolve(testData),
	})
)

beforeEach(() => {
	fetch.mockClear()
})

jest.mock('react-leaflet', () => {
	return {
		__esModule: true,
		MapContainer: jest.fn(({ children }) => {
			return (<>{children}</>)
		}),
		TileLayer: jest.fn(({ children }) => {
			return (<>{children}</>)
		}),
		Marker: jest.fn(({ children }) => {
			return (<>{children}</>)
		}),
		Popup: jest.fn(({ children }) => {
			return (<>{children}</>)
		}),
		useMapEvents: jest.fn(),
	}
})

jest.mock('react-leaflet-cluster', () => {
	return {
		__esModule: true,
		default: jest.fn(({ children }) => {
			return (<>{children}</>)
		}),
	}
})

jest.mock('next/font/google', () => {
	return {
		__esModule: true,
		Noto_Color_Emoji: jest.fn(() => {
			return {
				className: 'my-font',
			}
		}),
		Inter: () => ({
			className: 'inter-font-class'
		}),
	}
})

describe('<GeoFootballPage />', () => {
	it('renders the Geo-Football on Rails page', async () => {
		const { findByText, asFragment } = render(
			<Provider store={store}>
				<GeoFootballPage stationsByCountry={testClusters} />
			</Provider>
		)
		expect(await findByText('Geo-Football on Rails')).toBeInTheDocument()
		expect(asFragment()).toMatchSnapshot()
	})
})

describe('getServerSideProps', () => {
	it('fetches spatial data properly from API', async () => {
		const result = await getServerSideProps()

		expect(fetch).toHaveBeenCalledTimes(1)
		expect(result).toEqual({
			props: {
				stationsByCountry: testClusters
			}
		})
	})
})
