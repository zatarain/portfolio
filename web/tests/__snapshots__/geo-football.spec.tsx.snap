// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GeoFootballPage /> renders the Geo-Football on Rails page 1`] = `
<DocumentFragment>
  <div
    class="inter-font-class container geo-football"
  >
    <nav
      class="navigation-bar "
      id="navigation-bar"
      role="menubar"
    >
      <a
        class="logo"
        href="/"
      >
        <img
          alt="logo"
          data-nimg="1"
          decoding="async"
          height="48"
          loading="lazy"
          src="/logo.svg"
          style="color: transparent;"
          width="48"
        />
        Geo-Football on Rails
      </a>
      <ul
        class="sections"
      >
        <li>
          <a
            href="/#work-experience"
          >
            Experience
          </a>
        </li>
        <li>
          <a
            href="/#education"
          >
            Education
          </a>
        </li>
        <li>
          <a
            href="/#academic-projects"
          >
            Academic Projects
          </a>
        </li>
        <li>
          <a
            href="/#leadership"
          >
            Leadership
          </a>
        </li>
        <li
          class="action"
        >
          <a
            class="call-to-action"
            href="#"
          >
            Download
          </a>
        </li>
        <li
          class="dropdown"
        >
          <a
            href="#"
            role="menuitemcheckbox"
          >
            <i
              class="fa fa-bars"
            />
          </a>
        </li>
      </ul>
    </nav>
    <main>
      <form
        class="form"
        id="popup-add-form"
        method="POST"
      >
        <h3>
          Add New Marker
        </h3>
        <label
          for="name"
        >
          Name:
        </label>
        <input
          id="name"
          name="name"
          required=""
          role="textbox"
          type="text"
        />
        <div
          class="error"
        />
        <div
          class="location"
        >
          <div>
            <label
              for="latitude"
            >
              Latitude:
            </label>
            <input
              disabled=""
              id="latitude"
              name="latitude"
              type="text"
              value="51.47800"
            />
          </div>
          <div>
            <label
              for="longitude"
            >
              Longitude:
            </label>
            <input
              disabled=""
              id="longitude"
              name="longitude"
              type="text"
              value="0.00000"
            />
          </div>
        </div>
        <div
          class="country-time"
        >
          <div
            class="country"
          >
            <label
              for="country"
            >
              Country:
            </label>
            <select
              class="my-font"
              id="country"
              name="country"
            >
              <option
                class="my-font"
                value="GB"
              >
                ðŸ‡¬ðŸ‡§
              </option>
              <option
                class="my-font"
                value="IE"
              >
                ðŸ‡®ðŸ‡ª
              </option>
            </select>
          </div>
          <div>
            <label
              for="time-zone"
            >
              Timezone:
            </label>
            <input
              id="time-zone"
              name="time_zone"
              type="text"
            />
          </div>
        </div>
        <label
          for="info-en"
        >
          Additional information:
        </label>
        <input
          id="info-en"
          name="info_en"
          type="text"
        />
        <button
          type="submit"
        >
          Save
        </button>
      </form>
      <div
        class="popup"
        role="note"
      >
        <h3>
          Canary Wharf
        </h3>
        <dl>
          <dt>
            Country: 
          </dt>
          <dd>
            GB 
            <span
              class="my-font"
            >
              ðŸ‡¬ðŸ‡§
            </span>
          </dd>
          <dt>
            Position: 
          </dt>
          <dd>
            51.50361, -0.01861
          </dd>
          <dt>
            Timezone: 
          </dt>
          <dd>
            Europe/London
          </dd>
        </dl>
        <p>
          London Underground Station
        </p>
        <button
          class="delete"
        >
          <img
            alt="Delete"
            data-nimg="1"
            decoding="async"
            height="16"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fcdn-icons-png.flaticon.com%2F512%2F6711%2F6711573.png&w=32&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fcdn-icons-png.flaticon.com%2F512%2F6711%2F6711573.png&w=16&q=75 1x, /_next/image?url=https%3A%2F%2Fcdn-icons-png.flaticon.com%2F512%2F6711%2F6711573.png&w=32&q=75 2x"
            style="color: transparent;"
            width="16"
          />
          Delete
        </button>
      </div>
      <div
        class="popup"
        role="note"
      >
        <h3>
          Dublin Heuston
        </h3>
        <dl>
          <dt>
            Country: 
          </dt>
          <dd>
            IE 
            <span
              class="my-font"
            >
              ðŸ‡®ðŸ‡ª
            </span>
          </dd>
          <dt>
            Position: 
          </dt>
          <dd>
            53.34583, -6.29530
          </dd>
          <dt>
            Timezone: 
          </dt>
          <dd>
            Europe/Dublin
          </dd>
        </dl>
        <p>
          Irish Trains
        </p>
        <button
          class="delete"
        >
          <img
            alt="Delete"
            data-nimg="1"
            decoding="async"
            height="16"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fcdn-icons-png.flaticon.com%2F512%2F6711%2F6711573.png&w=32&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fcdn-icons-png.flaticon.com%2F512%2F6711%2F6711573.png&w=16&q=75 1x, /_next/image?url=https%3A%2F%2Fcdn-icons-png.flaticon.com%2F512%2F6711%2F6711573.png&w=32&q=75 2x"
            style="color: transparent;"
            width="16"
          />
          Delete
        </button>
      </div>
    </main>
    <footer
      class="footer"
    >
      <ul>
        <li>
          <h4>
            Tech Stuff
          </h4>
          <ul>
            <li>
              <a
                href="#"
              >
                C++
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                TypeScript
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Ruby
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Terraform
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Golang
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Python
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Cloud - AWS
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Git
              </a>
            </li>
          </ul>
        </li>
        <li>
          <h4>
            Skills
          </h4>
          <ul>
            <li>
              <a
                href="#"
              >
                Algorithms and Data Structures
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Mathematics
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Software Engineering
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Web Development
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Continuous Integration
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Tech Agnostic
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Proactivity
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Continuous Learning
              </a>
            </li>
          </ul>
        </li>
        <li>
          <h4>
            Personal Projects
          </h4>
          <ul>
            <li>
              <a
                href="#"
              >
                Algorists Group
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Unit Test Zeal
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Personal Website
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Jarnik Build System
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Spectral Clustering
              </a>
            </li>
          </ul>
        </li>
        <li>
          <h4>
            Volunteering
          </h4>
          <ul>
            <li>
              <a
                href="#"
              >
                LEGO Robotics and Tech for Kids
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Plant Trees
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                OmegaUp Latinoamerica
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Rider Photoshoot
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Charity Events
              </a>
            </li>
          </ul>
        </li>
        <li>
          <h4>
            Hobbies and fun
          </h4>
          <ul>
            <li>
              <a
                href="#"
              >
                Cooking
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Liverpool FC
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Anime
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Videogames
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Boardgames
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Watching sports
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Travel and Hiking
              </a>
            </li>
          </ul>
        </li>
        <li>
          <h4>
            Hates and boring
          </h4>
          <ul>
            <li>
              <a
                href="#"
              >
                Apple stuff
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Real Madrid FC
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Java stuff
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </footer>
  </div>
</DocumentFragment>
`;
